### PR TITLE
New Hcal Object - HcalQIEType 

### PR DESCRIPTION
**Pull request by walterjr**
First step to create a new object on the Data Base.

This PR only create the object (or class) for the next step (implementation on the code to create tags on the DB using the new object).

This DB container will provide per-HCAL-readout flag of the FE QIE electronics type for choosing a corresponding ADC coder in DIGI/RECO.

It involves the following packages:
CondFormats/DataRecord
CondFormats/HcalObjects

### CODE REVIEW

**Comment by abdoulline** [comment]
@diguida, @cermina, @franzoni, @ggovi, @mmusich, @davidlange6

we need this quickly revised/merged to proceed with HCAL HF dual readout DIGI/RECO updates for 2016...

**Comment by mmusich** [comment]
please test

**Comment by mmusich** [comment]
@walterjr @abdoulline  you need to edit  `CondCore/Utilities/src/CondDBFetch.cc` and ` CondCore/Utilities/src/CondDBImport.cc ` in order for the `conddb_import` tool to fetch and import the class `HcalQIEType`. This is needed for the dropbox to properly handle the payloads of this type.
You can have a look at the change needed here: https://github.com/cms-sw/cmssw/pull/9105/files

**Comment by abdoulline** [comment]
Walter,  please modify your branch the way Marco has asked:

CondCore/Utilities/src/CondDBFetch.cc
CondCore/Utilities/src/CondDBImport.cc
CondCore/Utilities/src/CondFormats.h 

to add HcalQIETypes (the same way - for instance - HcalRecoParams is added).
as in https://github.com/cms-sw/cmssw/pull/9105/files




https://github.com/mmusich/cmssw/blob/542a06584a0546f63dd9a5c2635e800dc258aa06/CondCore/Utilities/src/CondFormats.h#L187

**Comment by walterjr** [comment]
@mmusich @abdoulline 
At first moment we decide to divide the work in 3+1 steps (modifications+wait):
1) create the object - current PR
2) DB issues without change the RelVal test - a preliminary work can be found where [1] (including the modifications that you mention however will be edited for final version). Tests has been performed and it is working.
3) wait for the DB update the code
4) change the code to run the RelVal test - also on [1] however is necessary to remove some comments in some lines

For now is not necessary to change the CondCore/Utilities files because Step (1) do not take any "action" using the DB. All necessary changes to handle with the DB will be taken on Step (2).

[1] https://github.com/cms-sw/cmssw/compare/CMSSW_8_0_X...walterjr:NewHcalQIETypeNoRelVal

**Comment by abdoulline** [comment]
Walter,

(let Marco correct me, if I'm wrong) -
my understanding is that during the first current 1) step, DB people need 
to  modify DB to accomodate for the HcalQIETypes, so that one could 
upload HcalQIETypes to DB after it.
And for uploading to DB you need to move mentioned 3 files changes from 
your (prepared, next) step 2) to this one 1).

BTW, I presume you mean not RelVal, but runTheMatrix test in your 2)

S.

**Comment by abdoulline** [comment]
Walter, 
why can't you merge 1) + 2)  
I mean to add 2) to this branch? 
Do you see any problem, have  any objections? 

So that all the code changes would be submitted (without breaking runTheMatrix tests) at once?
Then DB experts can modify DB to accommodate for   
It would save time which would be needed for sequential approval of two PRs...

**Comment by walterjr** [comment]
Salavat,

Local test with (1)+(2) shows that the runTheMatrix test do not break.
Also is possible to create sqlite files using the new object and in addition also list and dump from the sqlite file is available.
If everybody agree I can update the branch and approval all in a single PR.

**Comment by abdoulline** [comment]
Walter,  
 
Once (1)+(2) will be approved/merged,  DB experts can proceed with preparing new table in DB for payload.
Then we will be able to upload whatever will be available (via Dropbox).
Please, update the branch.

**Comment by mmusich** [comment]
please test

**Comment by abdoulline** [comment]
Looking into logs I have the impression that IB used for jenkins tests 
cmssw-patch/CMSSW_8_0_X_2015-11-18-1100
is just broken (the same issue in all the faild tests):

File "/afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw-patch/CMSSW_8_0_X_2015-11-18-1100/python/DQMOffline/Configuration/DQMOffline_SecondStep_cff.py", line 16, in <module>
    from DQM.BeamMonitor.AlcaBeamMonitorClient_cff import *
  File "/afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw-patch/CMSSW_8_0_X_2015-11-18-1100/python/DQM/BeamMonitor/__init__.py", line 2, in <module>
    __path__.append(os.path.dirname(os.path.abspath(__file__).rsplit('/DQM/BeamMonitor/',1)[0])+'/cfipython/slc6_amd64_gcc493/DQM/BeamMonitor')
NameError: name 'os' is not defined

**Comment by mmusich** [comment]
let's try it again

**Comment by mmusich** [comment]
please test

**Comment by abdoulline** [comment]
@diguida, @cerminar, @franzoni, @ggovi, @mmusich 
please proceed with revision and approval...

**Comment by diguida** [comment]
@walterjr @abdoulline 
chiming in now.
Based on the previous discussion, my suggestion is the following:

1. making a PR which allows you to create, read and distribute your object (aka producer code). It must include also unit tests for the object: verifying that it is correctly stored (no matter what technology is used), and that the `EventSetup`is able to handle the record(s) delivering it.

2. making a new PR which allows to use the object in the HLT/DIGI/RECO steps (aka consumer code). There you must be sure that runTheMatrix is happy with your changes.

From a quick check to the PR, the requirements in 1. are accomplished. I am going through for more careful inspection.

**Comment by abdoulline** [comment]
@diguida, @franzoni, @ggovi, @mmusich

given that this simplest-possible single-integer-pre-readout DB object is mostly copy-paste of the existing (HcalRespCorrs) DB array, could you, please, consider merging it into the release to proceed with the DB changes for uploading test tag and then to proceed with aforementioned (by Salvatore) step 2...

**Comment by mmusich** [comment]
_1

File: CondCore/HcalPlugins/plugins/BuildFile [hunk 1]
```diff
@@ -175,4 +175,19 @@
...
+<use name=boost>
+<use name=boost_filesystem>
+<use name=boost_python>
+<use name=boost_regex>
+
+<use name=root>
+<use name=rootgraphics>
+
+<flags EDM_PLUGIN=1>
+</library>
```
Comment (diguida) [inline CondCore/HcalPlugins/plugins/BuildFile]: The `BuildFile` is now a clean XML file. You can drop this.

File: CondCore/HcalPlugins/plugins/HcalQIETypesPyWrapper.cc [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+			//MAIN FUNCTION:
+			datarepr.fillOneGain((*imageIter));
+
+			++(datarepr.nr);
+			++(datarepr.id);
+		}
+		return filename;
+	}
+}
+PYTHON_WRAPPER(HcalQIETypes,HcalQIETypes);
```
Comment (diguida) [inline CondCore/HcalPlugins/plugins/HcalQIETypesPyWrapper.cc]: The python payload inspector API here was supported only in V1.
We are working on a new implementation for V2.
So, please remove this file, and clean the BuildFile.

File: CondFormats/HcalObjects/xml/HcalQIETypes_default.xml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?xml version='1.0' encoding="UTF-8"?>
+<!DOCTYPE PoolDatabase SYSTEM "InMemory">
+ <PoolDatabase >
+  <PoolContainer name="HcalQIETypes" >
+   <Class name="HcalQIETypes"  mapping_version="HcalQIETypes_automatic_default" >
+   </Class >
+  </PoolContainer >
+ </PoolDatabase >
+
```
Comment (diguida) [inline CondFormats/HcalObjects/xml/HcalQIETypes_default.xml]: drop this, not needed.

**Comment by diguida** [comment]
@abdoulline @walterjr The +1 by @mmusich still holds. If you clean up according to what I proposed, even better.
Indeed, here you have everything you need:
* definition of the object and of the template specialisation for your general container
* definition of a record
* serialisation unit test
* unit tests for reading with ESSource and write via `PopCon`
* utilities for importing payloads.

File: CondCore/HcalPlugins/plugins/HcalQIETypesPyWrapper.cc (hunk 2 above)
Comment (walterjr) [inline CondCore/HcalPlugins/plugins/HcalQIETypesPyWrapper.cc]: In that case will be also necessary to remove the last lines of
CondCore/HcalPlugins/plugins/BuildFile.xml

Comment (diguida) [inline CondCore/HcalPlugins/plugins/HcalQIETypesPyWrapper.cc]: @walterjr true.
That's why I wrote:
> clean the BuildFile.
A correct statement is:
So, please remove this file, and clean the corresponding entry in `BuildFile.xml`.

**Comment by diguida** [comment]
please test

**Comment by diguida** [comment]
+1
* correct definition of a new condition object and of the template specialisation for the HCAL general container
* definition of a record associated to the object
* activation of serialisation unit test
* unit tests for reading with ESSource and write via `PopCon`
* utilities for importing payloads.

**Comment by mmusich** [comment]
ping @ggovi

**Comment by abdoulline** [comment]
@ggovi - could you, please, approve it...

**Comment by abdoulline** [comment]
@ggovi - could you, please, approve it ?
"Ceterum censeo..."(c)

**Comment by ggovi** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**